### ISSUE TITLE
Potential lock contention on TrieCache

### ISSUE THREAD

**Issue by bowenwang1996** [near/nearcore#10992]
After #10973, we still see long time spent on `ShardTries::update_cache`. There is an example of `update_cache` taking more than 2s after the node has been running for a while.
<img width="354" alt="Screen Shot 2024-04-08 at 3 44 13 PM" src="https://github.com/near/nearcore/assets/13601726/4ee017d7-1024-4060-9559-5222b0754dc4">

This seems to point to a lock contention on `TrieCache`. With the introduction of aggressive prefetching in #10935, there is potential contention when a thread tries to prefetch data and another thread calls `update_cache`. What is still unclear to me is how this can block for 2s. It might be related to the TrieCache implementation since it is not a simple LRU cache. This also happens quite rarely.

**Comment by bowenwang1996** [comment]
cc @Longarithm in case you have thoughts on this

**Comment by bowenwang1996** [comment]
We found a potential issue in https://github.com/near/nearcore/blob/4c0aa98cb7ecd5fd2a22c87788aa56d9223b7abb/core/store/src/trie/trie_storage.rs#L142. It is possible that value has size 0 (a [HashSet](https://docs.rs/near-sdk/5.1.0/near_sdk/collections/struct.LookupSet.html) for example). So this loop could potentially take a really long time. Another piece of evidence for this hypothesis is that in the trace linked above, `update_cache ` span has very little idle cpu time (3433 ns) and if it were due to lock contention, it should have much higher idle time

**Comment by bowenwang1996** [comment]
I instrumented the logic after acquiring locks and found that when it takes a long time, the time is indeed spent on the actual logic of updating cache, not acquiring the lock itself

**Comment by bowenwang1996** [comment]
@Longarithm could we simplify the logic by actually removing an element when we delete it from the cache instead of putting it into the deletion queue?

**Comment by nagisa** [comment]
> Another piece of evidence for this hypothesis is that in the trace linked above, update_cache span has very little idle cpu time (3433 ns) and if it were due to lock contention, it should have much higher idle time

The idle time metric is only really relevant for spans that are entered and/or exited multiple times, such as in asynchronous code (where each `poll` is an individual entry/exit.) In synchronous code the idle time will usually be quite low (even if/when waiting on a synchronous lock implementation,) as it usually enters the span immediately after it is created and only does so once until the span is destroyed.

In other words, it is not transparently measuring something like CPU time for us and cannot be used as a proxy to determine whether the code within the span was actually idle or not.

**Comment by bowenwang1996** [comment]
@nagisa unfortunately I was not able to get to the end of this issue. From what I see, it seems that the cache simply isn't performant enough. Tracing shows that each put/pop takes at least 1 microsecond and sometimes takes ~10 microseconds, which is quite slow. That may be due to the number of entries in the cache but could also due to tracing instrumentation (at 1μs latency I suspect tracing adds nontrivial overhead). We need to look more into it. If it is indeed the performance of the cache that is the issue, I think there may be several ways we can resolve it:

- Optimize the cache operations as much as possible
- Use some heavy lifting such as lock free LRU caches (looks like there is [a rust implementation](https://docs.rs/moka/latest/moka/))
- Do I understand correctly that for the correctness of Trie to work, we only need to force update cache on deletion but not necessarily insertion? It seems we mostly rely on trie nodes being fetched during execution anyways and not sure how much it helps to have it do a blocking update on writing trie changes. cc @Longarithm

**Comment by nagisa** [comment]
tracing is hundreds-of nanoseconds of overhead, but also its really taxing to use it to trace such small and frequent operations.

That said, one instrumentation that I added is the number of transaction ops so that we can see if there is a relationship between the number of ops and duration `commit` takes to execute. Looking some of the spans they take 700ms to execute even though there's only one transaction for them to process!

Then there's one that needs as much time for 17 transactions and another that needs to handle 12271 transactions (that makes it 50µs per transaction on average, but my suspicion is that we might have one transaction that's slow in there and plenty that are fast...).

**Comment by nagisa** [comment]
Early results with a pre-allocated `LruCache` seem to be promising. Will keep it running for a little while longer, see what things end up looking like after a few hrs and submit a PR.
